clients/drcachesim/tracer/instru_offline.cpp

```diff
@@ -597,8 +598,10 @@ offline_instru_t::instrument_instr(void *drcontext, void *tag, void **bb_field,
             return adjust;
         pc = dr_fragment_app_pc(tag);
     } else {
+        // This routine is called for the first instr in the expanded scatter/gather
+        // sequence, which turns out to be a non-app instr which doesn't have an app pc.
```

Comment:
It seems best to require that `app` be an app instr, rather than passing in flags and having special cases here.

Reply:
That might introduce some more special cases in tracer.cpp.

Here are the first few instrs in the expansion:
```
mov     %rax -> %gs:0x000000e8[8byte] // first instr in expansion. the reg spill.
lahf    -> %ah
seto    -> %al
<label>
mov    %rcx -> %gs:0x000000f0[8byte]
<label>
mov    %rdx -> %gs:0x000000f8[8byte]
<label>
kmovw  %k1 -> %ecx // first app instr in expansion.
test   %ecx $0x00000001
jz     @0x00007f7aed7b8060[8byte]
...
```
I assume you mean that we should just skip instrumenting the first `mov` (the reg spill), and instrument `kmovw` instead (which is an app instr). It is indeed a suitable candidate as it's guaranteed to execute in 0-iter cases too. But, this'd mean sacrificing the always-instrument-first-instr invariant (look for `drmgr_is_first_nonlabel_instr` in tracer.cpp), which may get a bit tricky. I can try.

Reply:
Right, that tracer logic is complex.  This also interacts with #4865 where I was assuming I'd have to tackle that same logic and try to refactor it.

Reply:
There does seem to be a gulf between the ideal of only app instrs being visible, or using first_app instead of first_nonlabel, and reality with messy sequences from app2app.

What if instead of passing in the scatter-gather the code instead checked for !instr_is_app and in that case it used the tag, for this PR: and assuming we can clean up the iteration in the tracer for #4865, maybe we can remove that later, but as it doesn't need extra bool params it feels like a cleaner less intrusive temporary (hopefully) solution.

Reply:
Removed the two extra bool params from `instrument_instr`. Switched to `!instr_is_app` to check whether the given bb has scatter/gather expanded. Added an assert check to verify that `!instr_is_app` iff `ud->scatter_gather`.

Reply:
Also, to find whether any repstr was expanded, reverted to the old check.

Reply:
Hmm.. found an issue when I fixed the assert statement to verify that `!app` happens if-and-only-if `scatter_gather`: `DR_ASSERT((ud->scatter_gather && !instr_is_app(instr)) ||
                  (!ud->scatter_gather && instr_is_app(instr)));`

Hitting the assert in `burst_malloc` for a non-app `nop` instr that is the first (and only) instruction in a bb.

Reply:
Why do we care whether !app coincides with scatter-gather?  I don't think we do.

Reply:
We are just assuming that !app corresponds to the first instr in the block, and we assume the app pc version of the tag corresponds to that.

Reply:
I was hoping that `!instr_is_app <=> ud->scatter_gather`: this is a stricter assumption that would've increased my confidence in the temporary workaround we're using. But it seems that `!instr_is_app` also happens in case of the `nop` at `replace_retaddr_sentinel`.

Anyway, I replaced the assert with: `instr_is_app || drmgr_is_first_nonlabel_instr` which should work for now. As discussed, for `!instr_is_app`, we get pc from the bb tag, and also decode pc to get the instr's size.